src/renderer/actions/repair.js

```diff
@@ -1,7 +1,6 @@
-
 import {progress, status} from "../stores/installation";
 import {remote} from "electron";
-import {promises as fs} from "fs";
+const fs = require("fs");
```

Comment:
Why go away from promises and also not use `import` syntax?

Reply:
This came from me not being a primary Javascript developer. I looked up the Electron docs and this is what came up in their examples. I even found some discussion which said that `require` is the better way.

But today I just looked it up again after seeing your review and now see that it's the other way around: `import` is the new, modern Javascript way. So I'll be converting all of these new imports to actual imports!

As for why I went away from promises: The code runs synchronously and used await, and I found a bunch of resources that said to use the Sync functions instead of awaiting promises, when you do native filesystem code. It's up to you, I can rewrite it back to using await-promises if you prefer that. Either one works basically the same way anyway.

Reply:
I/O using Promise/async/await is non-blocking on systems that support it (IOCompletionPort/io_uring) and thus generally goes easier on the hardware. The sync variants block the calling thread (think spin-lock).

Async I/O is always the better choice when your constraints allow you to use it, but it *does* require you to async/await the entire call chain due to language design...

Side-note: There's even a bug in node on windows when deleting directories using the sync calls (win32 RemoveDirectory), causing the directory to sometimes not having been deleted by the time the function returns.

Reply:
@Inve1951 I was there when Node was invented and was quite active in it when it was "the new hype", so yeah I am aware of the async revolution it brought (love Ryan Dahl's talk where he introduced Node).

But the async functions do the exact same work as the sync functions, but with extra memory and processing overhead. This works out to **the async functions being 4-7x slower than the synchronous functions** for serial tasks (where one file is written/read after the other, which is what BetterDiscord does).

The async functions are vital for Node servers where they must handle multiple clients. But they don't make much sense for desktop apps, since the GUI thread is totally separate and is handled by Electron's Chromium renderer, and the BetterDiscord Installer app itself is doing serial tasks which don't have any parallelism ([this is the reference I saw when I did the original changes](https://stackoverflow.com/questions/61032948/why-is-async-performance-worse-synchronous-on-blocking-file-i-o-in-node-js)). That's why I switched to Sync since it's much easier to write.

As for bugs in the synchronous functions, I highly doubt it since they're most likely fixed by now. I can't find that bug at all if I search for `node win32 RemoveDirectory sync bug` online.

But either way it doesn't matter. I'll switch all file I/O to async functions again even though there isn't any benefit for a desktop app. :)

Reply:
It seems you have more insight than you've lead with. :smile:
If there isn't actually a CPU hog problem using the sync calls and it even speeding things up then I don't see a reason to go back to Promises.
We do run that stuff in the renderer but I'm also aware that it doesn't actually update the UI until it completed it's I/O operations.
Either way not a big deal *IMO* since we are dealing with rather small files that even older hardware should be able to handle in a split second.

Reply:
@Inve1951 Thanks again for your additional review, I really appreciate it. :)

> It seems you have more insight than you've lead with. 😄

Hehe yeah, I used to be very active in JavaScript around 2009-2014 and made Node scripts and browser extensions. I just haven't been active in Node/JavaScript development for a few years, and have never made an Electron app myself. But since I am active in Rust and C++, it's not like I suddenly "forgot" everything about programming. :D It was easy to get back into JS. The hardest part was actually researching the "newest ways" to do things, since Node and JS have changed so rapidly that the internet is full of outdated answers and old documentation even from just a few years ago.

However, seeing how easy it is to build Electron apps in Node has inspired me to the possibility of using Electron in my own apps. The benefits of being able to make a web app that also runs on Desktop is awesome. I didn't think it was so easy to build cross-platform Electron apps, and it really impressed me.

> If there isn't actually a CPU hog problem using the sync calls and it even speeding things up then I don't see a reason to go back to Promises. We do run that stuff in the renderer but I'm also aware that it doesn't actually update the UI until it completed it's I/O operations.

Yeah, it was unnecessary to go back. But I had already gone back to Promises when this comment was written though. I haven't pushed the changes yet since I'm waiting for answers regarding the still-open questions/reviews here. There are only a few minor things left to discuss. After that I'll rebase (merge) all the new commits with the old ones to keep a clean pull request commit log. Hopefully a merge won't be far off after that.

> Either way not a big deal _IMO_ since we are dealing with rather small files that even older hardware should be able to handle in a split second.

That's exactly my feeling too. The files are small and are quickly written no matter what method we use. :)

Anyway, we're waiting for Zack to have time and energy to revisit this review. I managed to get in touch on Discord, and he's really busy with projects and real life etc (totally normal and understandable), so this may take a while. It's okay with me, and users can still use the build I provided at the top of this pull request. It's at 28 downloads now, so Linux users are being helped in the meantime. And it's still the same app, so that build is doing auto-update checks for new official releases at your official repo. :)

https://api.github.com/repos/Bananaman/BetterDiscord-Installer/releases/latest

Therefore it will be a smooth transition for people when there's a new official release. Although you'll need to give it a version number above 1.2.0 to make their clients detect a new update (sorry, I didn't think about that when I gave it a new version to distinguish it). ;)